browser/brave_stats/brave_stats_updater.cc

```diff
@@ -0,0 +1,375 @@
+  (void) MaybeDoThresholdPing(0);
+
+  // Log the full URL of the stats ping.
+  VLOG(1) << "Brave stats ping, url: " << final_url.spec();
+}
+
+void BraveStatsUpdater::OnThresholdLoaderComplete(
+    scoped_refptr<net::HttpResponseHeaders> headers) {
+  DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
+  GURL final_url = simple_url_loader_->GetFinalURL();
+  int response_code = -1;
+  if (headers)
+    response_code = headers->response_code();
+  if (simple_url_loader_->NetError() != net::OK || response_code < 200 ||
+      response_code > 299) {
+    VLOG(1) << "Failed to send threshold ping to update server"
+            << ", error: " << simple_url_loader_->NetError()
+            << ", response code: " << response_code
+            << ", url: " << final_url.spec();
+    return;
+  }
+
+  // Inform the client that the threshold ping completed, if requested.
+  if (!stats_threshold_callback_.is_null())
+    stats_threshold_callback_.Run(final_url);
+
+  // We only send this query once.
+  DisableThresholdPing();
+
+  // Log the full URL of the stats ping.
+  VLOG(1) << "Brave stats ping, url: " << final_url.spec();
+}
+
+void BraveStatsUpdater::OnServerPingTimerFired() {
+  // If we already pinged the stats server today, then we're done.
+  std::string today_ymd = brave_stats::GetDateAsYMD(base::Time::Now());
+  std::string last_check_ymd = pref_service_->GetString(kLastCheckYMD);
+  if (base::CompareCaseInsensitiveASCII(today_ymd, last_check_ymd) == 0)
+    return;
+
+  SendServerPing();
+}
+
+bool BraveStatsUpdater::IsReferralInitialized() {
+  return true;
+#if BUILDFLAG(ENABLE_BRAVE_REFERRALS)
+  return pref_service_->GetBoolean(kReferralInitialization) ||
+      pref_service_->GetBoolean(kReferralCheckedForPromoCodeFile);
+#else
+  return true;
+#endif
+}
+
+bool BraveStatsUpdater::HasDoneThresholdPing() {
+  return pref_service_->GetBoolean(kThresholdCheckMade);
+}
+
+void BraveStatsUpdater::DisableThresholdPing() {
+  pref_service_->SetBoolean(kThresholdCheckMade, true);
+  pref_service_->ClearPref(kThresholdQuery);
+}
+
+void BraveStatsUpdater::QueueServerPing() {
+  if (IsReferralInitialized()) {
+    StartServerPingStartupTimer();
+  } else {
+    pref_change_registrar_.reset(new PrefChangeRegistrar());
+    pref_change_registrar_->Init(pref_service_);
+    pref_change_registrar_->Add(
+        kReferralInitialization,
+        base::Bind(&BraveStatsUpdater::OnReferralInitialization,
+                   base::Unretained(this)));
+  }
+}  // namespace brave
+
+void BraveStatsUpdater::OnReferralInitialization() {
+  StartServerPingStartupTimer();
+}
+
+void BraveStatsUpdater::StartServerPingStartupTimer() {
+  server_ping_startup_timer_->Start(
+      FROM_HERE, base::TimeDelta::FromSeconds(kUpdateServerStartupPingDelay),
+      this, &BraveStatsUpdater::OnServerPingTimerFired);
+  DCHECK(server_ping_startup_timer_->IsRunning());
+}
+
+net::NetworkTrafficAnnotationTag
+    BraveStatsUpdater::AnonymousStatsAnnotation() {
```

Comment:
Does it really need to be a class method? I think this is just a free function that should live in the anomymous namespace